### ISSUE TITLE
Making a standalone YOLOv11 architecture. How to do it? (may apply to other Ultralytics YOLO series)

### ISSUE THREAD

**Issue by DWSuryo** [ultralytics/ultralytics#22795]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I want to make a standalone YOLOv11 model outside Ultralytics for image/video/camera detection only. I'm doing this because I want to explore to make a rather modular YOLOv11 algorithm and also make a further enhancement and improvements to YOLOv11.

First, I look at the `default.yaml` to see the hyperparameters, then look at `yolo11.yaml`. Here's what the default yolo11.yaml looks like:

```
# Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license

# Ultralytics YOLO11 object detection model with P3/8 - P5/32 outputs
# Model docs: https://docs.ultralytics.com/models/yolo11
# Task docs: https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 80 # number of classes
scales: # model compound scaling constants, i.e. 'model=yolo11n.yaml' will call yolo11.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.50, 0.25, 1024] # summary: 181 layers, 2624080 parameters, 2624064 gradients, 6.6 GFLOPs
  s: [0.50, 0.50, 1024] # summary: 181 layers, 9458752 parameters, 9458736 gradients, 21.7 GFLOPs
  m: [0.50, 1.00, 512] # summary: 231 layers, 20114688 parameters, 20114672 gradients, 68.5 GFLOPs
  l: [1.00, 1.00, 512] # summary: 357 layers, 25372160 parameters, 25372144 gradients, 87.6 GFLOPs
  x: [1.00, 1.50, 512] # summary: 357 layers, 56966176 parameters, 56966160 gradients, 196.0 GFLOPs

# YOLO11n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2/4
  - [-1, 2, C3k2, [256, False, 0.25]]
  - [-1, 1, Conv, [256, 3, 2]] # 3-P3/8
  - [-1, 2, C3k2, [512, False, 0.25]]
  - [-1, 1, Conv, [512, 3, 2]] # 5-P4/16
  - [-1, 2, C3k2, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]] # 7-P5/32
  - [-1, 2, C3k2, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]] # 9
  - [-1, 2, C2PSA, [1024]] # 10

# YOLO11n head
head:
  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 6], 1, Concat, [1]] # cat backbone P4
  - [-1, 2, C3k2, [512, False]] # 13

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 2, C3k2, [256, False]] # 16 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 13], 1, Concat, [1]] # cat head P4
  - [-1, 2, C3k2, [512, False]] # 19 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 10], 1, Concat, [1]] # cat head P5
  - [-1, 2, C3k2, [1024, True]] # 22 (P5/32-large)

  - [[16, 19, 22], 1, Detect, [nc]] # Detect(P3, P4, P5)

```

The backbone and head are crucial, especially at the modules. We got Conv, C3k2, SPPF, C2PSA, Concat, and Detect block. As I look to each block in `ultralytics/nn/modules/block.py` and `ultralytics/nn/modules/conv.py`, here's my understanding so far:

* **Conv** is a basic pytorch module, using Conv2D() and BatchNorm2D() with activation SiLU
* **Concat** is a basic pytorch concatenate cat() function
* **C3k2** inherits from C2f, then calls C3k. The C3k has Bottleneck class
  * **C2f** uses 2 convolutions with faster efficiency, which also uses Bottleneck class
  * The **C3k** inherits from C3 block, which has 3 convolutions with Bottleneck class
* **SPPF** is independent from SPP, uses 2 convolutions and concatenate later
* **C2PSA** the modified Position Sensitive Attention block, calls PSABlock
  * **PSABlock** calls the Attention class block
* **Detect** module is in `ultralytics/nn/modules/head.py`.

With these blocks, I think I can make a model. However, I want to make the model **in one model.py file** because I also want to export the model and the file to somewhere else. How do I connect each sequence in a backbone and head?

Another one I want to know is about how YOLOv11 Ultralytics do augmentations for dataset in `ultralytics/data/augment.py`. As I read, there are Mosaic, MixUp, and Letterbox, which latter will be used for validation if I'm not mistaken. For dataset loading, it is found in BaseDataset class from `ultralytics/data/base.py`.

This means I set my files into this:

* default.yaml (default hyperparameters)
* yolo11.yaml (model construction)
* coco.yaml (dataset initialization)
* dataset.py (for data loading)
* model.py
* augment.py
* plotting.py (for visualization)
* train.py
* val.py
* detect.py (for inference)

What I'm still struggling is to how to make a train.py and val.py. What I want to do is to run train.py and val.py separately so I can diagnose which one is wrong or error. I think that's what I'm asking so far. Unless there's someone who has already made a standalone version, maybe someone can help for this one.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @DWSuryo, thank you for your detailed question and for exploring YOLO11 in such depth 🚀  
This is an automated response to help you get oriented quickly — an Ultralytics engineer will also review your issue and assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, including model configuration, training, validation, and deployment workflows. Many architecture- and training-related questions are already covered there and may help guide your standalone design.

If this is a 🐛 Bug Report related to your custom standalone implementation or the current YOLO11 behavior, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE). This should include:
- A small, self-contained script (or set of scripts) demonstrating the issue  
- The exact `yolo11*.yaml` or custom model config you are using  
- Minimal sample data or clear instructions to reproduce the problem  

If this is a custom training or architecture ❓ Question (for example, about restructuring YOLO11 into a single `model.py`, custom `train.py` / `val.py`, or reusing `augment.py` and `BaseDataset`), please provide as much information as possible, such as:
- A short description or diagram of your desired standalone pipeline  
- Code snippets of your current `model.py`, `train.py`, and `val.py` structure  
- Any logs, errors, or mismatched behavior you’re seeing compared to the Ultralytics implementation  
And please verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), as these are often crucial when refactoring or modularizing the training/validation loops 💡

Join the Ultralytics community where it suits you best for additional discussion and peer help:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧  
- For in-depth, long-form discussions, see [Discourse](https://community.ultralytics.com/)  
- Or explore and share knowledge on our [Subreddit](https://www.reddit.com/r/Ultralytics/)  

## Upgrade

Before diving deep into debugging or refactoring, please ensure you are on the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This helps confirm that any behavior you are basing your standalone implementation on is up to date:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

These can be very helpful references while you are reverse-engineering or mirroring the training/validation pipeline in your own `train.py` and `val.py`.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit. This can serve as a reference for expected behavior while you design and validate your standalone implementation.

**Comment by glenn-jocher** [comment]
Nice breakdown of the blocks, your understanding is broadly in line with how YOLO11 is wired.  

The backbone/head “connections” are not hard‑coded in a `model.py`; they are built at runtime by `parse_model()` in `ultralytics/nn/tasks.py`, which reads each `[from, repeats, module, args]` row in `yolo11.yaml`, uses `from` to pick one or more previous layer outputs, and then uses `Concat` to merge multi‑input paths. If you want a single‑file `model.py`, the practical options are either to port that parser logic into your own project, or let Ultralytics build the model and then grab the plain PyTorch module as `YOLO("yolo11n.yaml").model` for use in your own code.  

For separate `train.py` / `val.py` you don’t need to re‑implement the training loop: you can keep all augmentations (`Mosaic`, `MixUp`, `LetterBox`, etc.) and dataloading exactly as Ultralytics does and just call the high‑level APIs, for example:  

```python
# train.py
from ultralytics import YOLO

model = YOLO("yolo11n.pt")  # or "yolo11n.yaml"
model.train(data="coco.yaml", epochs=100, imgsz=640)
```

```python
# val.py
from ultralytics import YOLO

model = YOLO("runs/detect/train/weights/best.pt")
model.val(data="coco.yaml", imgsz=640)
```

This keeps training and validation fully separated for debugging while still using the same dataset and augmentation stack implemented under `ultralytics/data`. If you decide to fully re‑implement everything, the [Model YAML Configuration Guide](https://docs.ultralytics.com/guides/model-yaml-config/) and the YOLO11 [detection task docs](https://docs.ultralytics.com/tasks/detect/) are the best references for how the graph and the train/val pipelines are assembled.

**Comment by DWSuryo** [comment]
Thanks for the reply. So far, my plan is to make `model.py`, `dataset.py`, and `loss.py` first. The `model.py` is of course to build the overall YOLOv11 architecture based on the blocks I've mentioned. The `dataset.py` will refer to YOLODataset class from `ultralytics\data\dataset.py` which inherits from BaseDataset at `ultralytics\data\base.py`.  Loss is a bit tricky here. As I read, YOLOv11 uses `v8DetectionLoss` which calls TAL class `TaskAlignedAssigner`. The TAL is in `ultralytics\utils\tal.py` so I copy most of the modules to my loss.py. That means the `loss.py` has necessary loss functions in `ultralytics\utils\loss.py` including TAL class.

Regarding training loop, for me I think I have to explore more on the training tasks and which modules should I pick. Personally, I would like to separate my `train.py` and `val.py` to diagnose every code I run, though I can make them run both at once later on. However, it's great you said that I might not need to reimplement a new `train.py` or 'val.py' if I'm assuming the program can run at once like how the Ultralytics version works.

I'll post my repo later

**Comment by glenn-jocher** [comment]
Sounds like a reasonable plan. One suggestion as you go: instead of duplicating `YOLODataset`, `v8DetectionLoss`, and `TaskAlignedAssigner` into your own `dataset.py` / `loss.py`, you may find it easier to first import them directly from `ultralytics.data` and `ultralytics.utils` and wire your custom `model.py` into a minimal train/val loop, then gradually replace or refactor pieces once everything is numerically matching the Ultralytics pipeline. When you publish the repo, feel free to drop a link back here; if you run into concrete issues, please include a small reproducible snippet plus the full traceback so we can help you pin down where train vs val is diverging.

**Comment by DWSuryo** [comment]
Here's the repo I've made: https://github.com/DWSuryo/YOLOv11-standalone/tree/faithful (use "faithful" branch. I have `config.py` to parse the `default.yaml` and `yolo11.yaml`.  Sure, it's a bit redundant but at this point it's still safe. There are separate diagnostic codes to see which image has to be fixed to avoid "corrupted" labels. I might clean up later.

Another one I'd like to tell is the parameters and FLOPs. Here's my result:
```
--- YOLOv11 Model Statistics ---
Scale      | Params (M)      | FLOPs (G)
---------------------------------------------
n          | 3.25            | 7.14
s          | 11.94           | 23.82
m          | 23.18           | 69.51
l          | 29.34           | 89.51
x          | 33.92           | 138.92
```

It's a bit different compared to official version. I noticed the significance in 'n' and 'x' scale. The mAP for scale 'n' for 5 epochs is okay, at least useable.

| epoch | train/box | train/cls | train/dfl | train/total | val/box | val/cls | val/dfl | val/total | metrics/precision | metrics/recall | metrics/mAP50 | metrics/mAP50-95 | lr |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 0 | 0.26907 | 7.00628 | 1.36849 | 7.57388 | 0.24745 | 6.10941 | 1.26452 | 6.80734 | 0.25347 | 0.08153 | 0.04594 | 0.02785 | 0.00802 |
| 1 | 0.22296 | 5.72197 | 1.17413 | 6.29438 | 0.21831 | 5.11843 | 1.13188 | 5.89434 | 0.26899 | 0.12825 | 0.09453 | 0.06315 | 0.00604 |
| 2 | 0.21158 | 5.30121 | 1.12186 | 5.92026 | 0.21504 | 4.93599 | 1.11919 | 5.75960 | 0.26572 | 0.15813 | 0.11883 | 0.07879 | 0.00406 |
| 3 | 0.19881 | 4.80996 | 1.06852 | 5.49882 | 0.20291 | 4.59319 | 1.06202 | 5.41148 | 0.32687 | 0.17477 | 0.14367 | 0.09711 | 0.00208 |
| 4 | 0.18964 | 4.45090 | 1.03011 | 5.19287 | 0.19074 | 3.99944 | 1.00548 | 4.93847 | 0.35209 | 0.23089 | 0.20352 | 0.14402 | 0.00010 |

I think this is the progress so far. I'll send updates later

**Comment by glenn-jocher** [comment]
Nice work getting a faithful version running and thanks for sharing the [YOLOv11-standalone repo](https://github.com/DWSuryo/YOLOv11-standalone/tree/faithful) and metrics.  

The params/FLOPs deltas you’re seeing (especially n/s/m/l being larger and x being much smaller than the reference) almost always come from small differences in how `scales` are applied in `parse_model()`—in particular the `depth`, `width`, `max_channels` tuple, the `make_divisible(..., 8)` rounding on `c2`, and the special handling of `C3k2` for `"mlx"` scales and other `base_modules`/`repeat_modules` logic described in the [tasks parse_model reference](https://docs.ultralytics.com/reference/nn/tasks/). A practical way to pin this down is to instantiate the official model and print a detailed summary, then compare layer‑by‑layer against your build, for example:  

```python
from ultralytics import YOLO

m = YOLO("yolo11n.yaml")
m.info(detailed=True)  # per-layer channels, params, FLOPs
```  

If after matching the per-layer shapes/params you still see a difference in FLOPs or training behaviour, feel free to post the specific layer comparison here and we can help you track down the remaining discrepancy.
